### ISSUE TITLE
[RFC] On-policy distillation (OPD) design compatible with uni-agent

### ISSUE THREAD

**Issue by wangtiance** [verl-project/verl#7998]
# On-Policy Distillation: Dual Teacher-Logprob Computation Paths

## Overview

OPD requires per-token teacher logprobs for every trained trajectory. Because
trajectories can originate from two different generation paths (verl's native
`AgentLoopWorker`, or an external agent framework such as [uni-agent's](https://github.com/verl-project/uni-agent/blob/main/uni_agent/framework/framework.py)
`AgentFramework`), the system supports two teacher-logprob computation
paths. They share all infrastructure and the loss consumption side; they differ
only in **who invokes the teacher, at what time, and how results are persisted**.

## Design Intent

The two paths mirror each other under different trajectory origins:

- **Path 1**: trajectories are produced inside verl (`AgentLoopWorker`) → teacher
  computation is attached at the production point, overlapping with training.
- **Path 2**: trajectories are produced outside verl ([uni-agent](https://github.com/verl-project/uni-agent) gateway) → the
  training step is the first point on verl's side where the complete batch is
  available. Attaching teacher computation there keeps the agent framework fully
  decoupled, structurally skips validation, avoids TQ schema extensions, and
  eliminates teacher compute on dropped trajectories.

The AgentLoopWorker rollout-time path is **retained**: it serves verl's native
OPD flows (v0 example, v1 e2e CI), costs nothing to keep, and removing it would
enlarge our diff against upstream verl.

## Shared Infrastructure (used by both paths)

| Component | Location | Responsibility |
|---|---|---|
| `MultiTeacherModelManager` | `verl/verl/trainer/ppo/v1/trainer_base.py` (`_setup`, step 8) | Runs in the trainer process; launches and manages teacher vLLM server replicas; exposes raw RPC clients via `get_client()` |
| `AsyncTeacherLLMServerManager` | `verl/verl/experimental/teacher_loop/teacher_manager.py` | Semantic layer: multi-teacher routing (`_resolve_teacher_key`), request construction (`max_tokens=1, prompt_logprobs=K`; K=0 for `k1` mode), response parsing/validation (`compute_teacher_logprobs_single`) |
| Distillation loss | `verl/verl/trainer/distillation/losses.py` | Consumes `teacher_ids` / `teacher_logprobs`; identical for both paths |

### Trigger conditions (all must hold)

1. `distillation.enabled=True` → trainer-side `use_teacher_policy=True`, teacher servers started.
2. Native verl agent loop in use (`agent_loop_manager_class` unset → v1 default `AgentLoopManagerTQ`).
3. Non-validation trajectory: per-sample gate in `_compute_teacher_logprobs`
   (`verl/verl/experimental/agent_loop/agent_loop.py`, `if self.distillation_enabled and not validate`).

### Call chain

```
main_ppo.py  TaskRunner.init_agent_loop_manager
  └─ agent_loop_manager_cls.create(teacher_client=trainer.get_teacher_client())
       └─ AgentLoopWorker.__init__ (agent_loop.py)
            └─ is_distillation_enabled → construct AsyncTeacherLLMServerManager

Per trajectory (agent_loop.py, AgentLoopWorker._run):
  run_agent_loop → _compute_score → _compute_teacher_logprobs
    ├─ routing_key = sample_kwargs[teacher_key]  (multi-teacher routing; auto for single teacher)
    ├─ teacher_server_manager.compute_teacher_logprobs_single(prompt_ids + response_ids)
    │    └─ teacher vLLM server: max_tokens=1, prompt_logprobs=K
    └─ results → output.extra_fields["teacher_ids" / "teacher_logprobs"]

Format & persistence:
  agent_loop.py: _pad_teacher_outputs → dense-pad to (prompt_width, response_width)
  ├─ v0: DataProto batch fields, flowing through the training pipeline
  └─ v1: AgentLoopManagerTQ writes to TransferQueue (per-token ragged fields)

Consumption:
  update_actor → compute_distillation_loss_reverse_kl_estimator
  (v1: no_padding_2_padding first asserts ragged length == prompt+response length)
```

### Motivation

uni-agent trajectories are produced by `GatewayAgentFramework` outside verl's
`AgentLoopWorker`, so Path 1's call site never fires. Computing teacher logprobs
at train time keeps the agent framework fully OPD-agnostic and eliminates an
entire class of custom TQ schema extensions (ragged promotion, padding-template
fixes, `is_validate` context injection) plus wasted teacher compute on
trajectories that are later dropped (staleness/filtering).

### Architecture (Option A)

```text
┌────────────────────┐
│ uni-agent gateway  │  trajectories: prompts / responses only
│ (OPD-agnostic)     │  (no teacher fields, no is_validate gate)
└─────────┬──────────┘
          ▼
   ┌──────────────┐
   │ TransferQueue │
   └──────┬───────┘
          │ KVBatchMeta(keys, tags)
          ▼
┌──────────────────────────────────────┐
│ Trainer — _step_once                 │
│  _balance_batch (upsample + pad)     │
│  _compute_old_log_prob               │
│  advantage                           │
│  _compute_teacher_log_prob  ← NEW    │
│   (step 8.5; gated by                │
│    compute_at_train_time=True)       │
│    │ read prompts/responses from TQ  │
│    ▼                                 │
│  AsyncTeacherLLMServerManager        │
│    │ RPC: max_tokens=1,              │
│    ▼      prompt_logprobs=K          │
│  teacher vLLM servers                │
│    (MultiTeacherModelManager)        │
│    │ teacher_ids / teacher_logprobs  │
│    ▼ write back to TQ (jagged; after │
│      upsample → lengths always match)│
│  _update_actor ──► distillation loss │
└──────────────────────────────────────┘
```

### Trigger conditions (all must hold)

1. `distillation.enabled=True` (same as Path 1).
2. `distillation.compute_at_train_time=True` (**new switch, default False**).
3. Batch enters `_step_once` — validation goes through `_validate` instead, so
   skipping teacher computation is **structural**, no per-sample gate needed.

### Call chain

```
trainer_base.py: _setup step 8 (shared) + new:
  self.teacher_server_manager = AsyncTeacherLLMServerManager(
      config=self.config, teacher_client=self.teacher_model_manager.get_client())

Per training step (trainer_base.py: _step_once):
  step 3  _balance_batch (includes upsample_batch_to_divisible_size;
           synthetic padding samples already in place)
  step 4  _compute_old_log_prob
  ...
  step 8.5 [NEW] _compute_teacher_log_prob(batch)             ← gated by the switch
    ├─ tq.kv_batch_get(keys, select_fields=["prompts", "responses"])
    │    └─ nested tensors; per-sample lengths via .offsets().diff()
    ├─ per sample: torch.cat([prompts[i], responses[i]]) → sequence_ids
    ├─ _run_async_in_temp_loop(asyncio.gather(
    │      *[teacher_server_manager.compute_teacher_logprobs_single(ids) ...]))
    │    (multi-teacher: routing_key from batch.tags[teacher_key]; single teacher: auto)
    └─ torch.nested.as_nested_tensor(..., layout=torch.jagged)
         + tq.kv_batch_put(fields=tu.get_tensordict(...))   ← mirrors _compute_reward_colocate
  step 9  _update_actor (unchanged)

Why the read side needs zero changes:
  replay_buffer.py returns KVBatchMeta(partition_id, keys, tags) with fields=None
  → workers fetch ALL fields per key from TQ (transferqueue_utils.py)
  → teacher fields written at train time are indistinguishable from
    rollout-time-written ones on the consumption side.
```

Because the write happens **after** upsampling, synthetic padding samples also
receive teacher tensors of exactly the synthetic sequence length (128), making
length mismatches impossible by construction.

## Path 2: Modification Points in verl

All changes are confined to the v1 trainer and the distillation config schema.
The rollout side (agent loop), the loss path, and the TQ read path are
untouched.

1. **`verl/verl/trainer/ppo/v1/trainer_base.py`** — three edits:
   - **`_setup`** (after step 8 teacher-manager init): construct the semantic
     layer once, reusing the existing clients:
     ```python
     self.teacher_server_manager = (
         AsyncTeacherLLMServerManager(config=self.config,
                                      teacher_client=self.teacher_model_manager.get_client())
         if self.use_teacher_policy else None
     )
     ```
   - **New method `_compute_teacher_log_prob(batch, metrics)`** (~60 lines),
     modeled on the existing `_compute_reward_colocate` (same "read TQ → compute
     → write back to TQ" pattern, same jagged-nested write format as
     `rm_scores`).
   - **`_step_once`**: insert step 8.5 between advantage computation and
     `_update_actor`, gated by
     `self.use_teacher_policy and self.config.distillation.get("compute_at_train_time", False)`.
2. **`verl/verl/workers/config/distillation.py`** — add one field to
   `DistillationConfig`:
   `compute_at_train_time: bool = False` (plus docstring).
   A minimal implementation could read the key via OmegaConf `.get()` without a
   schema change; the RFC recommends the proper dataclass field for
   discoverability and validation.

**Explicit non-changes:**

- `verl/verl/experimental/agent_loop/agent_loop.py` — untouched; Path 1 remains
  the default for native flows.
- Loss path (`verl/verl/trainer/distillation/`) — untouched.
- TQ read path / replay buffer — untouched (`KVBatchMeta.fields=None` ⇒ workers
  fetch all fields per key).
- `verl/verl/trainer/ppo/padding_utils.py` — unchanged. Its template builder
  never sees teacher fields under Path 2 (they are written after upsampling).

**Follow-up (optional, not part of the initial version):** run step 8.5
concurrently with `_compute_old_log_prob` in a thread (independent: async RPC
from the trainer process vs. Ray worker-group call) to hide teacher prefill
latency. v0-trainer support is a non-goal.

## Alternatives Considered: Framework-Side Rollout-Time Computation (Option B, Rejected)

The first implementation mirrored Path 1 inside uni-agent: teacher logprobs were
computed at trajectory finalization time by a postprocessor and persisted to TQ
through a custom ragged-field schema:

```text
┌────────────────────────────────────────────────┐
│ uni-agent gateway                              │
│  trajectory generation                         │
│    ▼                                           │
│  _TeacherLogprobPostprocessor   ← requires:    │
│    │  • teacher_client instance injection      │
│    │  • is_validate context injection          │
│    ▼                                           │
│  AsyncTeacherLLMServerManager                  │
│   (re-implemented in recipe — drift risk)      │
│    │ RPC: max_tokens=1, prompt_logprobs=K      │
│    ▼                                           │
│  teacher vLLM servers                          │
│    │ teacher_ids / teacher_logprobs            │
│    ▼ extra_fields                              │
└────┼───────────────────────────────────────────┘
     ▼ custom nested ragged TQ schema
┌──────────────┐   padding template must rebuild
│ TransferQueue │   custom per-token fields — fragile
└──────┬───────┘
       ▼
  Trainer: _update_actor ──► distillation loss
```

We implemented this and rejected it as the long-term design:

1. **Duplicates verl semantics** — the framework must re-implement teacher
   routing, the `k1`/`topk` request protocol, and server-lifecycle handling
   that `AsyncTeacherLLMServerManager` already owns; the two copies drift as
   verl evolves.
2. **Fragile TQ schema extensions** — custom per-token fields need nested
   ragged registration, write-time length validation, and awareness in the
   synthetic-padding template. We hit both failure modes in production
   (`ragged_idx` misconfiguration corrupting TQ writes; padding
   length-mismatch assertion in `no_padding_2_padding`).
3. **Validation gating leaks** — rollout-side scoring also fires for
   validation trajectories, so `is_validate` must be threaded through generic
   postprocessor injection; train-time computation skips validation
   structurally.
4. **Wasted teacher FLOPs** — rollout-time scores trajectories later dropped
   by staleness/filtering; train-time scores only consumed batches.
5. **Not generalizable** — a train-side hook in verl serves any external
   trajectory producer; the framework-side approach is re-implemented per
   framework.

Its two real advantages — teacher prefill fully overlapped with training, and
per-sample routing kwargs naturally available at generation time — are
recovered in Path 2 via the concurrency follow-up and `batch.tags[teacher_key]`
routing.

## Mutual Exclusivity and Safety

| Scenario | Path 1 | Path 2 | Outcome |
|---|---|---|---|
| Native agent loop, flag off (default) | ✅ | — | Status quo |
| uni-agent (`AgentFrameworkRolloutAdapter`), flag on | — (trajectories bypass `AgentLoopWorker`) | ✅ | Target state |
| uni-agent, flag off | — | — | Loss-side missing-field error; misconfiguration surfaces fast |
| Native + flag on | ✅ | ✅ | Double compute (Path 1 has no off switch); misconfiguration |

The root of mutual exclusivity: `agent_loop_manager_class` is a single class per
run — trajectories either pass through `AgentLoopWorker` or are written to TQ
directly. **No legal configuration triggers both paths.**

## Notes

1. **Multi-teacher routing**: Path 1 derives the routing key per sample from
   `sample_kwargs[teacher_key]`; Path 2 uses `batch.tags[teacher_key]` (tags
   already carry `data_source`; precedent in `trainer_base.py` validation-metric
   aggregation). Capability is equivalent.
2. **Performance**: Path 2 inserts teacher prefill serially before
   `update_actor`. It can be hidden by running it concurrently with
   `_compute_old_log_prob` (the two are independent: one is async RPC from the
   trainer process, the other a Ray worker-group call). Marked as a follow-up
   optimization, not required for the initial version.
3. **Config**: enable Path 2 with `+distillation.compute_at_train_time=True`.
   When enabled, the rollout-time postprocessor chain
   (`_TeacherLogprobPostprocessor`, TQ field promotion, `is_validate` injection)
   should be removed from the recipe.
